File: kernel_spark/src/main/scala/io/delta/DeltaCatalog.scala
package io.delta

import io.delta.kernel.Operation
import org.apache.hadoop.conf.Configuration
import org.apache.spark.sql.connector.catalog._
import org.apache.spark.sql.connector.expressions.{IdentityTransform, NamedReference, Transform}
import org.apache.spark.sql.types.StructType
import org.apache.spark.sql.util.CaseInsensitiveStringMap

import java.util
import java.util.UUID

import scala.collection.JavaConverters._

class DeltaCatalog extends TableCatalog {
  import DeltaCatalog._

  private var catalogName: String = _
  private lazy val engine =
    io.delta.kernel.defaults.engine.DefaultEngine.create(new Configuration())

  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
    this.catalogName = name
  }

  override def listTables(namespace: Array[String]): Array[Identifier] = {
    Array.empty
  }

  override def loadTable(ident: Identifier): Table = {
    new DeltaTable(ident.name())
  }

  override def createTable(
      ident: Identifier,
      schema: StructType,
      partitions: Array[Transform],
      properties: util.Map[String, String]): Table = {
    val path = if (properties.containsKey("path")) {
      properties.get("path")
    } else {
      s"/tmp/table_${UUID.randomUUID().toString.replace("-", "").substring(0, 4)}"
    }

    logger.info(
      s"createTable: ident=$ident, schema=$schema, " +
        s"partitions=${partitions.mkString("Array(", ", ", ")")}, properties=$properties, " +
        s"path=$path")

    val partitionCols = partitions.map(extractPartitionColumn)

    val result = io.delta.kernel.Table
      .forPath(engine, path)
      .createTransactionBuilder(engine, "kernel-spark-dsv2", Operation.CREATE_TABLE)
      .withSchema(engine, SchemaUtils.convertSparkSchemaToKernelSchema(schema))
      .withPartitionColumns(engine, partitionCols.toList.asJava)
      .build(engine)
      .commit(engine, io.delta.kernel.utils.CloseableIterable.emptyIterable())

    logger.info(s"createTable: resultVersion=${result.getVersion}")

    val table = new DeltaTable(path)
    inMemoryTables.put(ident.name(), table)
    table
  }

  override def dropTable(ident: Identifier): Boolean = {
    if (inMemoryTables.contains(ident.name())) {
      inMemoryTables.remove(ident.name())
      true
    } else {
      false
    }
  }

  override def renameTable(oldIdent: Identifier, newIdent: Identifier): Unit = {
    throw new UnsupportedOperationException("renameTable is not supported")
  }

  override def alterTable(ident: Identifier, changes: TableChange*): Table = {
    throw new UnsupportedOperationException("alterTable is not supported")
  }

  override def tableExists(ident: Identifier): Boolean = {
    inMemoryTables.contains(ident.name())
  }

  override def name(): String = catalogName

  private def extractPartitionColumn(transform: Transform): String = {
    logger.info(s"transform: $transform")
    // Check if the transform is an identity transform
    if (transform.name() == "identity" && transform.references().nonEmpty) {
      // Get the first reference, which should be the column
      transform.references()(0) match {
        case namedRef: NamedReference =>
          logger.info(s"namedRef: $namedRef")
          logger.info(
            s"namedRef.fieldNames: ${namedRef.fieldNames().mkString("Array(", ", ", ")")}")
          namedRef.fieldNames().mkString(".")
        case _ => throw new RuntimeException("bad aa")
      }
    } else {
      throw new RuntimeException("bad bb")
    }
  }
}

object DeltaCatalog {
  val logger = org.slf4j.LoggerFactory.getLogger(this.getClass)

  // identifier -> table
  val inMemoryTables = scala.collection.mutable.Map[String, DeltaTable]()
}

File: kernel_iceberg_spark/src/main/java/io/delta/DeltaCatalog.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package io.delta;

import io.delta.kernel.Operation;
import io.delta.kernel.defaults.engine.DefaultEngine;
import io.delta.kernel.engine.Engine;
import io.delta.kernel.utils.CloseableIterable;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import org.apache.hadoop.conf.Configuration;
import org.apache.iceberg.CatalogProperties;
import org.apache.iceberg.CatalogUtil;
import org.apache.iceberg.PartitionField;
import org.apache.iceberg.PartitionSpec;
import org.apache.iceberg.Schema;
import org.apache.iceberg.Table;
import org.apache.iceberg.catalog.Catalog;
import org.apache.iceberg.catalog.Namespace;
import org.apache.iceberg.catalog.TableIdentifier;
import org.apache.iceberg.exceptions.NoSuchTableException;
import org.apache.iceberg.hadoop.Configurable;
import org.apache.iceberg.hadoop.HadoopFileIO;
import org.apache.iceberg.io.FileIO;
import org.apache.iceberg.relocated.com.google.common.base.Joiner;
import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
import org.apache.iceberg.util.LocationUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DeltaCatalog implements Catalog, Configurable<Configuration> {

  private static final Logger LOG = LoggerFactory.getLogger(DeltaCatalog.class);
  private static final String HIVE_WAREHOUSE_PROP = "hive.metastore.warehouse.dir";
  private static final Joiner SLASH = Joiner.on("/");

  private Configuration conf = null;
  private String name = null;
  private Map<String, String> catalogProperties = null;
  private String warehouse = null;
  private FileIO fileIO = null;

  @Override
  public String name() {
    return name;
  }

  @Override
  public List<TableIdentifier> listTables(Namespace namespace) {
    return Collections.emptyList();
  }

  @Override
  public boolean dropTable(TableIdentifier identifier, boolean purge) {
    throw new UnsupportedOperationException("Drop is not supported");
  }

  @Override
  public void renameTable(TableIdentifier from, TableIdentifier to) {
    throw new UnsupportedOperationException("Rename is not supported");
  }

  @Override
  public Table createTable(
      TableIdentifier identifier,
      Schema schema,
      PartitionSpec spec,
      String location,
      Map<String, String> properties) {
    LOG.info(
        "createTable ::: identifier {} with schema {} and spec {} and location {} and properties {}",
        identifier,
        schema,
        spec,
        location,
        properties);
    return createTableHelper(identifier, schema, Optional.of(spec));
  }

  @Override
  public Table createTable(
      TableIdentifier identifier,
      Schema schema,
      PartitionSpec spec,
      Map<String, String> properties) {
    LOG.info(
        "createTable ::: identifier {} with schema {} and spec {} and properties {}",
        identifier,
        schema,
        spec,
        properties);
    return createTableHelper(identifier, schema, Optional.of(spec));
  }

  @Override
  public Table createTable(TableIdentifier identifier, Schema schema, PartitionSpec spec) {
    LOG.info("createTable ::: identifier {} with schema {} and spec {}", identifier, schema, spec);
    return createTableHelper(identifier, schema, Optional.of(spec));
  }

  @Override
  public Table createTable(TableIdentifier identifier, Schema schema) {
    LOG.info("createTable ::: identifier {} with schema {}", identifier, schema);

    return createTableHelper(identifier, schema, Optional.empty());
  }

  private Table createTableHelper(
      TableIdentifier identifier, Schema schema, Optional<PartitionSpec> partitionSpecOpt) {
    String tableLocation = tableLocation(identifier);
    Engine engine = DefaultEngine.create(conf);
    io.delta.kernel.Table kernelTable = io.delta.kernel.Table.forPath(engine, tableLocation);
    io.delta.kernel.TransactionBuilder txnBuilder =
        kernelTable
            .createTransactionBuilder(engine, "iceberg", Operation.CREATE_TABLE)
            .withSchema(engine, SchemaUtils.fromIcebergSchema(schema.asStruct()));

    if (partitionSpecOpt.isPresent()) {
      final List<String> partSpecNames =
          partitionSpecOpt.get().fields().stream()
              .map(PartitionField::name)
              .collect(Collectors.toList());
      txnBuilder = txnBuilder.withPartitionColumns(engine, partSpecNames);

      LOG.info("Scott > createTableHelper :: partSpecNames {}", partSpecNames);
    }

    txnBuilder.build(engine).commit(engine, CloseableIterable.emptyIterable());

    LOG.info("createTable ::: Created table at path {}", tableLocation);

    return new DeltaTable(identifier, conf, tableLocation);
  }

  @Override
  public TableBuilder buildTable(TableIdentifier identifier, Schema schema) {
    LOG.info("Scott > DeltaCatalog :: buildTable() :: identifier {}", identifier);
    return new DeltaTableBuilder(identifier, schema, conf);
  }

  private String tableLocation(TableIdentifier ident) {
    return SLASH.join(warehouse, ident.name());
  }

  @Override
  public Table loadTable(TableIdentifier ident) {
    String tableLocation = tableLocation(ident);

    LOG.info("loadTable ::: ident {}, location {}", ident, tableLocation);
    try {
      return new DeltaTable(ident, conf, tableLocation);
    } catch (io.delta.kernel.exceptions.TableNotFoundException ex) {
      LOG.info("loadTable :: TABLE NOT FOUND");
      throw new NoSuchTableException("Table not found: " + ident, ex);
    }
  }

  @Override
  public void initialize(String name, Map<String, String> properties) {
    LOG.info("Initializing DeltaCatalog with name: {}", name);

    if (null == conf) {
      LOG.warn("No Hadoop Configuration was set, using the default environment Configuration");
      conf = new Configuration();
    }

    this.name = name;
    this.catalogProperties = ImmutableMap.copyOf(properties);

    if (properties.containsKey(CatalogProperties.WAREHOUSE_LOCATION)) {
      this.warehouse =
          LocationUtil.stripTrailingSlash(properties.get(CatalogProperties.WAREHOUSE_LOCATION));
      conf.set(HIVE_WAREHOUSE_PROP, warehouse); // keep the Configuration in sync
    } else {
      this.warehouse = LocationUtil.stripTrailingSlash(conf.get(HIVE_WAREHOUSE_PROP, null));
    }

    Preconditions.checkArgument(
        warehouse != null, "Missing required property: %s", CatalogProperties.WAREHOUSE_LOCATION);

    String ioImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
    this.fileIO =
        ioImpl == null ? new HadoopFileIO(conf) : CatalogUtil.loadFileIO(ioImpl, properties, conf);
  }

  @Override
  public void setConf(Configuration conf) {
    this.conf = conf;
  }
}
